I2P integration

HarshSawarkar opened the pull request:
This PR includes a new` README.md` for testing the I2P and Tor functionality in Bisq. Both I2P and Tor integrations have been tested. Although there is some startup lag due to differences in initialization times between I2P and Tor, either network is typically sufficient to launch the Bisq desktop app.

**I2P functionality has been verified for:**

- Offer creation

- Group messaging

- Everything is working smoothly.

Note: I2P routers are designed for long-term operation. For users who are less active or start the app occasionally, the I2P startup time may be longer. 

**Remaining tasks:**

- Fix parallel initialization of I2P and Tor

- Address a few remaining minor UI.

I’m open to suggestions and will be happy to address any review comments.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full I2P transport support: config, seed nodes, proxy lists, key management, address resolution, peer online checks, dynamic I2P proxy selection.
  * Platform-native system standby prevention for Windows, macOS, and Linux.

* **Improvements**
  * Better embedded I2P router configuration, certificate handling, and async router/session management.
  * Increased permitted I2P address length; safer key bundle handling.

* **Bug Fixes**
  * Filtered null seed addresses; I2P addresses exempted from ban checks.

* **Documentation**
  * Added I2P usage and build/run guide.

* **Chores**
  * Updated I2P and JNA dependencies; added SSL certificates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HenrikJannsen commented:
Can you address the coderabbitai complaints and response and resolve them?
Can you square the 3 commits or at least the 2 with same commit message?
Directory `bisq2_Alice_tor` was accidentally checked it.

HenrikJannsen commented:
`defaultI2pPrivateKey = ""` need to be added to all configs like for seed node

HenrikJannsen commented:
`import bisq.dto.security.keys.I2pKeyPairDto;` was missing in `DtoMappings`.


I get following error when running a seed node without external router. 

```
ERROR - Cannot find I2P installation in /Users/dev/Library/Application Support/bisq2_seed_i2p/i2p/i2p-dir-base - Will probably be just a router with no apps or console at all!
WARNING: Unable to read /proc/cpuinfo
java.io.FileNotFoundException: /proc/cpuinfo (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:213)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:152)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:106)
	at net.i2p.util.NativeBigInteger.getCPUInfo(NativeBigInteger.java:1246)
	at net.i2p.util.NativeBigInteger.resolveCPUType(NativeBigInteger.java:444)
	at net.i2p.util.NativeBigInteger.<clinit>(NativeBigInteger.java:349)
	at net.i2p.crypto.CryptoConstants.<clinit>(CryptoConstants.java:48)
	at net.i2p.crypto.ElGamalEngine.<clinit>(ElGamalEngine.java:63)
	at net.i2p.I2PAppContext.initializeElGamalEngine(I2PAppContext.java:695)
	at net.i2p.I2PAppContext.elGamalEngine(I2PAppContext.java:688)
	at net.i2p.router.Router.warmupCrypto(Router.java:1376)
	at net.i2p.router.Router.startupStuff(Router.java:442)
	at net.i2p.router.Router.runRouter(Router.java:685)
	at bisq.network.i2p.I2pEmbeddedRouter.startEmbeddedRouter(I2pEmbeddedRouter.java:180)
	at bisq.network.i2p.I2pEmbeddedRouter.<init>(I2pEmbeddedRouter.java:74)
	at bisq.network.i2p.I2pEmbeddedRouter.getI2pEmbeddedRouter(I2pEmbeddedRouter.java:49)
	at bisq.network.p2p.node.transport.I2PTransportService.initialize(I2PTransportService.java:150)
	at bisq.network.p2p.ServiceNode.getInitializedDefaultNode(ServiceNode.java:270)
```

HenrikJannsen commented:
When running with external router I can get seed node running, but at desktop I get error:
```
14:07:56.430 ERROR [Network.IO-0.ServiceNodesByTransport.getInitializedDefaultNode-I2P] b.d.DesktopApplicationService: Error at networkFuture.initialize java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "net.i2p.I2PAppContext.isRouterContext()" because "context" is null
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run$$$capture(CompletableFuture.java:1770)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "net.i2p.I2PAppContext.isRouterContext()" because "context" is null
	at net.i2p.client.naming.NamingService.createInstance(NamingService.java:755)
	at bisq.network.i2p.util.I2PNameResolver.<init>(I2PNameResolver.java:41)
	at bisq.network.i2p.I2pClient.<init>(I2pClient.java:56)
	at bisq.network.i2p.I2pClient.lambda$getI2pClient$0(I2pClient.java:64)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at bisq.network.i2p.I2pClient.getI2pClient(I2pClient.java:63)
	at bisq.network.p2p.node.transport.I2PTransportService.getClient(I2PTransportService.java:195)
	at bisq.network.p2p.node.transport.I2PTransportService.initialize(I2PTransportService.java:165)
	at bisq.network.p2p.ServiceNode.getInitializedDefaultNode(ServiceNode.java:270)
	at bisq.network.p2p.ServiceNodesByTransport.lambda$getInitializedDefaultNodeByTransport$1(ServiceNodesByTransport.java:142)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run$$$capture(CompletableFuture.java:1768)
	... 4 common frames omitted
	```

HarshSawarkar commented on network/network/src/main/java/bisq/network/p2p/services/peer_group/PeerGroupService.java:
```diff
@@ -243,7 +243,7 @@ public boolean isNotBanned(Peer peer) {
     }
 
     public boolean isNotBanned(Address address) {
-        return banList.isNotBanned(address);
+        return true;
```
This was pushed by mistake; reverting the changes

HenrikJannsen commented:
Can you address the review complaints from coderabbitai?
The dev data directory is still checked in

HenrikJannsen commented:
Can you resolve the coderabbitai complaints and resolved the merge conflicts?

HarshSawarkar commented on security/src/main/proto/security.proto:
```diff
@@ -49,15 +49,13 @@ message TorKeyPair {
-  bytes privateKey = 1;
-  bytes publicKey = 2;
-  string destination = 3;
+  bytes destinationKey = 1;
 }
```
Not needed — those fields were never used in practice, so wire compatibility isn't a concern

HarshSawarkar commented on network/i2p/src/main/java/bisq/network/i2p/I2pClient.java:
```diff
@@ -17,247 +17,218 @@
-                } else {
-                    //Else we will use an embedded router
-                    //Is this needed???
-                    manager = I2PSocketManagerFactory.createManager(privKeyInputStream); // Blocking while router builds tunnels
-                    if (manager == null) {
-                        log.info("No I2P router found, initializing embedded one ...");
-                        manager = i2pRouter.getManager(privKeyFile);
-                    }
-                }
+            // Wrap those bytes in a ByteArrayInputStream and pass it to createDisconnectedManager:
+            I2PSocketManager manager;
+            try (ByteArrayInputStream privKeyStream = new ByteArrayInputStream(rawDestBytes)) {
+                manager = I2PSocketManagerFactory.createDisconnectedManager(
+                        privKeyStream,
+                        host,
+                        port,
+                        null
+                );
+            } catch (I2PSessionException e) {
+                throw new RuntimeException(e);
             }
```
The destination already encapsulates the public/private keys and address, so using it directly is correct.

HarshSawarkar replied:
The PrivateKeyFile format is exactly what getDestination() returns — it includes both public and private key material. I've tested and verified this behavior.

HarshSawarkar commented on network/i2p/src/main/java/bisq/network/i2p/I2pEmbeddedRouter.java:
```diff
@@ -195,84 +189,288 @@ private void startEmbeddedRouter(boolean extendedI2pLogging) throws IOException
+
+    public boolean isPeerOnline(Destination destination) {
+        return routerContext.commSystem().wasUnreachable(destination.getHash());
+    }
```
Fixed

HarshSawarkar commented on network/i2p/src/main/java/bisq/network/i2p/I2pEmbeddedRouter.java:
```diff
@@ -82,68 +115,24 @@ private I2pEmbeddedRouter(String i2pDirPath, int inboundKBytesPerSecond, int out
         log.trace("I2P Router Status changed to: {}", i2pRouterStatus.name());
-        if(i2pRouterStatus != latestStatus) {
+        if (i2pRouterStatus != latestStatus) {
             // Status changed
```
Fixed

HarshSawarkar commented on apps/desktop/desktop/src/main/java/bisq/desktop/main/content/user/user_profile/create/step2/CreateNewProfileStep2View.java:
```diff
@@ -140,7 +142,13 @@ protected void onViewAttached() {
+            } catch (GeneralSecurityException e) {
+                throw new RuntimeException(e);
+            }
+        });
```
fixed

HarshSawarkar commented on network/network/src/main/java/bisq/network/http/HttpClientsByTransport.java:
```diff
@@ -43,13 +49,33 @@ public BaseHttpClient getHttpClient(String url,
+                Proxy proxy = selectWorkingProxy(i2pConfig.getProxyList());
+                yield new ClearNetHttpClient(url, userAgent, proxy);
+            }
             case CLEAR -> new ClearNetHttpClient(url, userAgent);
```
Fixed

HarshSawarkar commented on network/network/src/main/java/bisq/network/NetworkService.java:
```diff
@@ -130,8 +131,11 @@ public NetworkService(NetworkServiceConfig config,
         networkIdService = new NetworkIdService(persistenceService, keyBundleService, supportedTransportTypes, defaultPortByTransportType);
-        httpClientsByTransport = new HttpClientsByTransport();
+        httpClientsByTransport = new HttpClientsByTransport(i2pCfg);
 
```
Fixed

HarshSawarkar commented on network/network/src/main/java/bisq/network/http/HttpClientsByTransport.java:
```diff
@@ -43,13 +49,45 @@ public BaseHttpClient getHttpClient(String url,
+        int size = proxyList.size();
+        for (int i = 0; i < size; i++) {
+            int idx = Math.abs(counter.getAndIncrement()) % size;
+            I2PTransportService.ProxyEndpoint candidate = proxyList.get(idx);
```
Fixed

HenrikJannsen commented:
Can you rebase on main? There are some conficts.

HarshSawarkar commented:
> On OSX PreventSleepService does not recognize the jni import in IntelliJ. I can though run it via gradle.
> 
> OSX Alice works. When starting Bob on same machine i get error:
> 
> ```
> 14:04:15.125 INFO  [Network.IO-0.ServiceNodesByTransport.getInitializedDefaultNode-I2P] b.n.i.I2pEmbeddedRouter: Launching I2P Router... 
> Starting I2P 2.7.0-0
> ERROR - Cannot find I2P installation in /Users/dev/.bisq2/services/bisq.network.i2p.I2pEmbeddedRouter - Will probably be just a router with no apps or console at all!
> 14:04:15.132 INFO  [JavaFX Application Thread] b.d.DesktopView: Attaching view to stage took 204 ms 
> WARN: There may be another router already running. Waiting a while to be sure...
> ERROR: There appears to be another router already running!
>        Please make sure to shut down old instances before starting up
>        a new one.  If you are positive that no other instance is running,
>        please delete the file /Users/dev/.bisq2/services/bisq.network.i2p.I2pEmbeddedRouter/pid/router.ping
> 14:05:20.242 INFO  [Thread-0.DesktopExecutable.shutdownHook] b.d.DesktopApplicationService: shutdown 
> 14:05:20.243 INFO  [ForkJoinPool.commonPool-worker-2] b.h.HttpApiService: shutdown 
> ```
> 
> On Linux I cannot get the app started. Tried various times also with shutting down the OSX app, but it failes with: `Reseed failed - check network connection`. It might be that my internetprovider makes issues but I use same wifi for both.

To run a second embedded router, you first need to delete the router.ping file using:
`rm /home/user/.bisq2/services/bisq.network.i2p.I2pEmbeddedRouter/pid/router.ping
`
After that, running two embedded routers simultaneously is possible.

For an external router, you don’t need to delete the router.ping file — you can connect both Alice and Bob smoothly.

Could you please share your logs for Linux? I think this issue might be related to your network configuration or environment, because I’m able to run it on my local Linux machine as well as on the VPS where the code is hosted (which is also Linux).

HarshSawarkar commented:
> but it failes with: `Reseed failed - check network connection`

The reseeding process may take some time to complete, so it’s best to allow it to finish, as reseeding happens only once.

HarshSawarkar commented on network/i2p/src/main/java/bisq/network/i2p/I2pClient.java:
```diff
@@ -17,247 +17,244 @@
 
-        return sessionMap.get(sessionId);
-    }
+        synchronized (sessionId.intern()) {
```
Updated

HarshSawarkar commented on network/network/src/main/java/bisq/network/http/HttpClientsByTransport.java:
```diff
@@ -43,13 +66,45 @@ public BaseHttpClient getHttpClient(String url,
+            }
+        }
+        throw new RuntimeException("No reachable I2P proxy available");
+    }
```
Updated

HarshSawarkar commented:
> Can you address the AI code review issues?

I have addressed all the AI code review issues except https://github.com/bisq-network/bisq2/pull/3454#discussion_r2150693156 [https://github.com/bisq-network/bisq2/pull/3454#discussion_r2150693156 ](url)

To address this issue, I’ve modified the method to initialize transports in parallel. This allows the faster transport to publish its addresses and establish a connection immediately, while the other continues its setup in the background. I’ve tested it: both Tor and I2P can connect, but the second transport doesn’t always publish its addresses on the desk screen, even though the peer connections are successful. Do you have any suggestions for ensuring that the slower transport also publishes its addresses?

HarshSawarkar commented:
Thanks, @axpoems. The VPS seed nodes were down somehow, so your seed node was not able to connect to them. Could you please check now?

HarshSawarkar commented:
Also, @axpoems , please ignore the README.md in this PR as it is outdated. I will remove it. Just follow the usual methods for running the seed node and desktop app.

HarshSawarkar commented:
> I tested again and it works well when I use an external i2p router. However, I was not able to make it work without, I keep getting this with no success:
> 
> ```
> Error reseeding: java.io.IOException: unknown signer: xxx@mail.i2p for content type: reseed
> Reseed got no router infos from https://i2p.xxx.in/i2pseeds.su3?netid=2
> Reseed got 0 router infos from https://i2p.xxx.in/i2pseeds.su3?netid=2 with 1 errors
> Reseeding from https://reseed.memcpy.io/i2pseeds.su3?netid=2
> ```

Thanks @axpoems for your time and effort. Manually reseeding the embedded router by copying the reseed ZIP into the I2P data directory helps with firewall and censorship resistance, but it’s not scalable for new users. I’m looking for a better solution.

HenrikJannsen commented:
Hallo @HarshSawarkar Any updates on the issues with the embedded router?

HenrikJannsen commented:
> > Hallo @HarshSawarkar Any updates on the issues with the embedded router?
> 
> Hi @HenrikJannsen ,
> 
> It took a while to hear back from the I2P maintainer, but they’ve recently responded. Attached is a screenshot of their review of the EmbeddedRouter class, where I raised concerns about the reseed process being flaky on some systems while working fine on others.

From what I remember when I shared the logs last time you said that the seed servers have been the same as on your node.

Tested now again on OSX / M3 and same issue. 
Maybe its some native lib which is embedded?  But last time it was also not working on my Ubuntu machine.

Here the relevant logs:

```
ERROR - Cannot find I2P installation in /Users/dev/Library/Application Support/bisq2_Alice_i2p/i2p - Will probably be just a router with no apps or console at all!
Warning - No client apps or router console configured - we are just a router
Reseed start
Reseeding from https://reseed.i2pgit.org/i2pseeds.su3?netid=2
Reseed got no router infos from https://reseed.i2pgit.org/i2pseeds.su3?netid=2
Reseeding from https://www2.mk16.de/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: i2p-reseed@mk16.de for content type: reseed
Reseed got no router infos from https://www2.mk16.de/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://www2.mk16.de/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://reseed.onion.im/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: lazygravy@mail.i2p for content type: reseed
Reseed got no router infos from https://reseed.onion.im/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://reseed.onion.im/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://reseed-pl.i2pd.xyz/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: r4sas-reseed@mail.i2p for content type: reseed
Reseed got no router infos from https://reseed-pl.i2pd.xyz/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://reseed-pl.i2pd.xyz/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://reseed.stormycloud.org/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: admin@stormycloud.org for content type: reseed
Reseed got no router infos from https://reseed.stormycloud.org/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://reseed.stormycloud.org/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://i2p.novg.net/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: igor@novg.net for content type: reseed
Reseed got no router infos from https://i2p.novg.net/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://i2p.novg.net/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://reseed.memcpy.io/i2pseeds.su3?netid=2
Reseed got no router infos from https://reseed.memcpy.io/i2pseeds.su3?netid=2
Reseeding from https://reseed2.i2p.net/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: echelon3@mail.i2p for content type: reseed
Reseed got no router infos from https://reseed2.i2p.net/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://reseed2.i2p.net/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://reseed.diva.exchange/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: reseed@diva.exchange for content type: reseed
Reseed got no router infos from https://reseed.diva.exchange/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://reseed.diva.exchange/i2pseeds.su3?netid=2 with 1 errors
Reseeding from https://i2pseed.creativecowpat.net:8443/i2pseeds.su3?netid=2
Reseed got no router infos from https://i2pseed.creativecowpat.net:8443/i2pseeds.su3?netid=2
Reseeding from https://reseed-fr.i2pd.xyz/i2pseeds.su3?netid=2
Reseed got no router infos from https://reseed-fr.i2pd.xyz/i2pseeds.su3?netid=2
Reseeding from https://banana.incognet.io/i2pseeds.su3?netid=2
Reseed got no router infos from https://banana.incognet.io/i2pseeds.su3?netid=2
Reseeding from https://i2p.ghativega.in/i2pseeds.su3?netid=2
Error reseeding: java.io.IOException: unknown signer: arnavbhatt288@mail.i2p for content type: reseed
Reseed got no router infos from https://i2p.ghativega.in/i2pseeds.su3?netid=2
Reseed got 0 router infos from https://i2p.ghativega.in/i2pseeds.su3?netid=2 with 1 errors
Reseed failed - check network connection
Ensure that nothing blocks outbound HTTP or HTTPS, check the logs, and if nothing helps, read the FAQ about reseeding manually.
Consider enabling a proxy for https on the reseed configuration page
```

HenrikJannsen commented:
@wodoro Could you give it a try?

HarshSawarkar commented:
Thanks, @wodoro , for fixing the issue. I’ve merged the changes, and hopefully all systems will now be able to reseed successfully.
@HenrikJannsen , could you please test it once?

 I’d also appreciate your review and insights on other parts of the PR. One such area is `I2pClient`, do you notice anything off there as well?

wodoro commented:
I still had issues with getting it to run
thought I made some changes and seem to be able to get it to run, will send another PR for it soon

wodoro commented:
No, reseed is successful. the problem is somewhere else.

wodoro commented:
The reason I did not immediately make a PR was that I wasn't sure what I did next was really a fix, I wanted to confirm it,
and after testing more it seems it really wasn't. so right now i2p starts up, but gets stuck at connecting to p2p network, which you said needs vps nodes to be running.

everything should work after the seed nodes are running then I guess

HarshSawarkar commented:
> The reason I did not immediately make a second PR was that I wasn't sure what I did next was really a fix, I wanted to confirm it, and after testing more it seems it really wasn't. so right now i2p starts up (after the previous PR), but gets stuck at connecting to p2p network, which you said needs vps nodes to be running.
> 
> everything should work after the seed nodes are running then I guess
> 
> edit: I managed to successfully launch application, I guess nodes are up now ?

yes the vps is up now. Are you able to connect to the seed nodes?

HenrikJannsen commented:
I got it working with a fresh data dir. 

Then after restart it took 5 sec to get I2P started but then very slow to connect to p2p network. it showed `I2pEmbeddedRouter: Not connected to I2P Network.` in the logs. But finally got started up but PeerExchange keeps failing as well as inventory requests. 

A 3rd start took 47 sec to get I2P started and same as before, connections but no data.

Another restart with fresh data dir:
starting i2p: 54 sec
connect to peers:59 sec
after 2.5 min data loaded.

wodoro commented:
@HarshSawarkar Let's get this ready to merge, please start by rebasing to main and addressing bot suggestions and resolving them, reply to the bot in case it's incorrect and then resolve it, so the bot gets updated
after that I'll start reviewing the PR and we can address problems and missing features together so we can get this in

wodoro commented:
what do you mean by abrupt restarts? stopping and starting the app in IDE for testing for example ?
I don't see any correlation
It's possible that the other end of the connection is just also stopping suddenly
please let me know if you encounter this again, I'll investigate it
